Add scaler implementation for liiklus

ericbottard opened the pull request:
This PR adds an implementation of a scaler for liiklus (https://github.com/bsideup/liiklus).

It is greatly inspired by the Kafka implementation and comes with some basic integration tests.

I had some issues with `dep` (would never complete resolution) and had to add the missing `Empty` protobuf definition by hand. Feel free to solve otherwise.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/kedacore/keda?pullRequest=289) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/kedacore/keda?pullRequest=289) before we can accept your contribution.<br/><br/>:x: ericbottard [sign now](https://cla.opensource.microsoft.com/kedacore/keda?pullRequest=289)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/kedacore/keda?pullRequest=289) it.</sub>

ahmelsayed commented:
for `dep` issue, you might need to install `hg` see this https://github.com/kedacore/keda/issues/255

ericbottard commented:
@ahmelsayed That fixed it, thanks!

bsideup commented on hack/LiiklusService.proto:
```diff
@@ -0,0 +1,139 @@
+
+option go_package = "liiklus";
+
+service LiiklusService {
```
I'm pretty sure you can include only the rpc endpoints you need for the integration, to reduce the size of the generated code and self-document the usage

ericbottard replied:
It's far easier to just grab the latest protobuf definition for liiklus and not worry about what is needed and what is not when the file will be updated IMO.

bsideup commented on pkg/scalers/liiklus_scaler.go:
```diff
@@ -0,0 +1,167 @@
+	var totalLag uint64 = 0
+	ctx1, cancel1 := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel1()
+	gor, err := s.client.GetOffsets(ctx1, &liiklus_service.GetOffsetsRequest{
```
FYI since `groupVersion` parameter is omitted, it will always use the first version (v0). 
I would suggest to add `groupVersion` to `liiklusMetadata` for those who use it

ericbottard replied:
Added support in commit 3372b5c

jeffhollan commented:
Any updates on this? Who has the next step?

ahmelsayed commented:
@bsideup has a couple of comments on the PR. He probably knows liiklus better than I do :) and the CLA.

From my side I wanted to test it, but I haven't had time to check liiklus yet.

bsideup commented:
Since Liiklus in an experimental phase atm, it does not have enough documentation, so I will just post a quick-starter with Docker here:
```
$ docker run -it --rm \
    -e storage_positions_type=MEMORY \
    -e storage_records_type=MEMORY \
    -p 8080:8080 \
    -p 6565:6565 \
    -p 8081:8081 \
    bsideup/liiklus:0.9.0-rc2
```

It will run an instance of Liiklus with in-memory positions (aka offsets) and records storages, and will listen for http traffic (e.g. `/health` or `/prometheus`) on port 8080, for gRPC on port 6565 and RSocket on port 8081.

Note that starting with 0.9.0-rc1 it supports gRPC's reflection service (very convenient for publishing from CLI)

ericbottard commented:
Hi, here are some updates:
- signed the CLA
- rebased on top of current master with a re-run of `dep ensure` (had to clean `vendor/` to work around [this issue](https://github.com/golang/dep/issues/1407)
- added support for `groupVersion`
- force pushed the whole thing (I have a copy of previous state if needed)

I believe this PR is ready for review/merge in its current feature set. Happy to open additional PRs to refine after that.

ericbottard commented:
Hi,

any updates on this ? Happy to help move it forward.

yaron2 commented:
Can you resolve the conflicts?

ericbottard commented:
Force pushed the 3 relevant topics on top of current master

yaron2 commented:
Is it possible to use grpc-mock for tests instead of spinning up a real server?

https://github.com/grpc/grpc-go/blob/master/Documentation/gomock-example.md

ericbottard commented:
> Is it possible to use gomock for tests instead of spinning up a real server?

I'm sure I can do that :) Thanks for the pointer!

ericbottard commented:
> Is it possible to use gomock for tests instead of spinning up a real server?

Please see 512517c

ericbottard commented:
@yaron2 May I bump this thread for a new review ? I believe it is in mergeable state, let me know if anything else can be done (including work in a follow up PR). Thanks!

ahmelsayed commented:
I finally managed to setup everything and create a simple producer/consumer and all worked great! 
Thank you @ericbottard and @bsideup! and sorry for the delay on the PR.

bsideup commented:
@ahmelsayed great news, thank you!

ericbottard commented:
No worries, thanks for your time!
